TorchProphet

karl-richter opened the pull request:
Initial version of a Prophet wrapper for NeuralProphet.
The wrapper aims to reduce the effort of migrating from Prophet to NeuralProphet by setting default arguments Prophet-like and slightly changing the behaviour of some functions. The ultimate target is to allow any application using Prophet to experiment with / migrate to NeuralProphet by changing only one line:

```
from prophet import Prophet
```
to
```
from neuralprophet import TorchProphet as Prophet
```

In the notebook under `tutorials/feature-use/prophet.ipynb`, the official [Prophet quick-start tutorial](https://facebook.github.io/prophet/docs/quick_start.html#python-api) demonstrates the migration from Prophet to NeuralProphet using the wrapper. 

We tested compatibility of the Prophet wrapper using all tutorial notebooks provided by Prophet ([Link](https://github.com/facebook/prophet/tree/main/notebooks)). The table indicates to which degree the Prophet wrapper supports the migration by changing only one line.

| Prophet notebook  | Status  |  Comment  |
| ------------- | ------------- | ------------- |
| `additional_topics.ipynb`  | Not supported  | Additional features, specific to Prophet |
| `diagnostics.ipynb`  | Not supported  | Exploration of model components, specific to Prophet |
| `multiplicative_seasonality.ipynb`  | Supported  |  |
| `non-daily_data.ipynb`  | Supported  | Only minor plots are not supported |
| `outliers.ipynb`  | Supported  |  |
| `quick_start.ipynb`  | Supported  |  |
| `saturating_forecasts.ipynb`  | Not supported  | Saturating forecasts are not supported by NeuralProphet |
| `seasonality,_holiday_effects,_and_regressors.ipynb`  | Partly supported  | Events and holidays supported, Conditional seasonality not supported  |
| `trend_changepoints.ipynb`  | Supported  | Functionality supported, plots partly not supported |
| `uncertainty_intervals.ipynb`  | Supported  | Uncertainty intervals are supported, Bayesian sampling not |

ourownstory commented on neuralprophet/wrapper.py:
```diff
@@ -1,8 +1,73 @@
+        # holidays, seasonality_prior_scale, holidays_prior_scale,
+        # changepoint_prior_scale, mcmc_samples, interval_width, uncertainty_samples, stan_backend
+        # Set Prophet-like default args
+        kwargs["quantiles"] = [0.9, 0.1]
```
you can map `interval_width` to `prediction_interval`, see https://github.com/ourownstory/neural_prophet/pull/715/

ourownstory reviewed (commented):
check on imploications of __init__
from .TorchProphet import plot_plotly, plot_components_plotly

karl-richter commented:
> check on imploications of **init** from .TorchProphet import plot_plotly, plot_components_plotly

It looks like there are no implications since NeuralProphet binds all plotting to the forecaster (eg. `NeuralProphet(...).plot(...)`). I anyways updated the tutorial notebook and removed the import from `__init__` since I agree that it might become confusing in the future.

ourownstory commented:
@karl-richter how did you resolve the plotting utils distinction issue?

karl-richter commented:
> @karl-richter how did you resolve the plotting utils distinction issue?

@ourownstory I removed the imports in `__init__` and just import directly from the TorchProphet file. I also changed this in the tutorial notebook.

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+            self.events_df.drop(["lower_window", "upper_window"], axis=1, errors="ignore", inplace=True)
+
+        # Overwrite NeuralProphet properties
+        self.name = "Prophet"
```
maybe better TorchProphet

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+        """
+        Validates the inputs to Prophet.
+        """
+        raise NotImplementedError("Not required in NeuralProphet as all inputs are automatically checked.")
```
instead of an error, we could just log an error and move on, as this should not cause any issues?

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+    ):
+        # Check for unsupported features
+        if seasonality_prior_scale or holidays_prior_scale or changepoint_prior_scale:
+            log.info(
```
maybe raise these logs to an error level?

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+
+        This function is not supported in NeuralProphet.
+        """
+        raise NotImplementedError(
```
instead of an error, we could just log an error and move on, as this should not cause any issues?

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+    changepoint_prior_scale: Not supported for regularisation in NeuralProphet,
+        please use the `_reg` args instead.
+    mcmc_samples: Not required for NeuralProphet
+    interval_width: Float, width of the uncertainty intervals provided
```
update description for NP: mention these are symmetrically converted to quantiles and point towards that interface being an alternative.

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+
+    Parameters
+    ----------
+    growth: String 'linear', 'logistic' or 'flat' to specify a linear, logistic or
```
remove logistic

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+    Parameters
+    ----------
+    growth: String 'linear', 'logistic' or 'flat' to specify a linear, logistic or
+        flat trend.
```
note: address 'flat' being called 'off' in NP: convert this if given

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+        that holiday.
+    seasonality_mode: 'additive' (default) or 'multiplicative'.
+    seasonality_prior_scale: Not supported for regularisation in NeuralProphet,
+        please use the `_reg` args instead.
```
please write out the "_reg" arg name

ourownstory replied:
(multiple)

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+            weekly_seasonality=weekly_seasonality,
+            daily_seasonality=daily_seasonality,
+            seasonality_mode=seasonality_mode,
+            # TODO: activate once PR merged prediction_interval=interval_width,
```
still relevant?

karl-richter replied:
Yes, looks like. There is still no `prediction_interval` in forecaster.py on main.

karl-richter replied:
I re-added the workaround with converting the interval_width to quantiles if no quantiles are provided.

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+        )
+
+    def fit(self, df, **kwargs):
+        """Fit the Prophet model.
```
Please adopt NP docstring as far as reasonable

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+        return metrics_df
+
+    def predict(self, df=None, **kwargs):
+        """Predict using the prophet model.
```
Please adopt NP docstring as far as reasonable

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+        return df
+
+    def predict_trend(self, df):
+        """Predict trend using the prophet model.
```
Please adopt NP docstring as far as reasonable

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+        return df_future
+
+    def add_seasonality(self, name, period, fourier_order, prior_scale=None, mode=None, condition_name=None, **kwargs):
+        """Add a seasonal component with specified period, number of Fourier
```
Please adopt NP docstring as far as reasonable

ourownstory replied:
/ remove descriptions of parameters not supported

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+        return super(TorchProphet, self).add_seasonality(name, period, fourier_order, **kwargs)
+
+    def add_regressor(self, name, prior_scale=None, standardize="auto", mode="additive", **kwargs):
+        """Add an additional regressor to be used for fitting and predicting.
```
Please adopt NP docstring as far as reasonable / remove descriptions of parameters not supported

ourownstory commented on neuralprophet/TorchProphet.py:
```diff
@@ -0,0 +1,553 @@
+
+        Returns
+        -------
+        The prophet object.
```
TorchProphet

ourownstory reviewed (changes requested):
Maybe move the tutorial to a new folder 'guides' in tutorials?

ourownstory commented:
Re: Tutorial:
- name file Prophet_to_TorchProphet
- Disclaimer: ... It allows forecasting practitioners using Prophet to quickly evaluate the feasibility of adopting NeuralProphet.
- in section "Identical behaviour": in 'Future regressors' and 'Holidays' you seem to use a few times TorchProphet when you likely  meant to say NeuralProphet.
- the form of the `interval_width` arg should be identical in Prophet and TorchProphet
- in Neural Network architecture: no hidden layer is default.
- in the basic plot: why does the NP plot not show uncertainty?

karl-richter reviewed (commented):
Adressed comments

karl-richter commented:
> Re: Tutorial:
> 
> * name file Prophet_to_TorchProphet
> * Disclaimer: ... It allows forecasting practitioners using Prophet to quickly evaluate the feasibility of adopting NeuralProphet.
> * in section "Identical behaviour": in 'Future regressors' and 'Holidays' you seem to use a few times TorchProphet when you likely  meant to say NeuralProphet.
> * the form of the `interval_width` arg should be identical in Prophet and TorchProphet
> * in Neural Network architecture: no hidden layer is default.
> * in the basic plot: why does the NP plot not show uncertainty?

Adressed all comments in the notebook and TorchProphet. I re-added the logic to convert the interval to quantiles and with that also the plot should contain uncertainty.
